Expand README.md to cover almost everything (#713)
go-debos/debos: doc/man/debos.md

enough rope to do whatever tweaking which might be required behind the
scenes.

> debos expects a YAML file as input, called a recipe. The recipe is

Reviewer: ```suggestion
debos expects a YAML file, called a recipe, as it's input. The recipe is
```

> pre-processed by a templating engine and then describes a list of actions

Reviewer: ```suggestion
pre-processed by the Golang templating engine and describes a list of actions
```

which debos runs sequentially. Actions are self-contained and independent of
each other.

> The recipe file format and templating engine are described in the *Recipe

Reviewer: link to `Recipe syntax` seciton ?

> The recipe file format and templating engine are described in the *Recipe

Reviewer: `The recipe syntax and ...`

> syntax* section. Every available action, with its properties and a worked

Reviewer: ```suggestion
syntax* section. Each available action, with its properties and a worked
```

> example, is documented in the *Actions* section. The variables debos exports

Reviewer: link to actions section ?

> to the build environment and the `origin` property are described in the

Reviewer: ```suggestion
to the build environment and the value for the `origin` property are described in the
```

> *Environment variables* section.

Reviewer: Link?

The reference documentation in this README is generated by hand and may lag
behind the code. The authoritative, automatically generated action

> documentation is published at

Reviewer: ```suggestion
documentation is published on the
```

> [godoc.org/github.com/go-debos/debos/actions](https://godoc.org/github.com/go-debos/debos/actions);

Reviewer: ```suggestion
[godoc documentation](https://godoc.org/github.com/go-debos/debos/actions).
```

> when in doubt, treat it as the more up-to-date source.

Reviewer: ```suggestion
When in doubt, treat the [godoc documentation](https://godoc.org/github.com/go-debos/debos/actions) as the more up-to-date source.
```

debos -t image:"debian-arm64.tgz" example.yaml
```

# RECIPE SYNTAX

A recipe is a YAML file. Before it is parsed as YAML it is run through the

> Go [text/template](https://pkg.go.dev/text/template) engine, so a recipe is

Reviewer: ```suggestion
Golang [template engine](https://pkg.go.dev/text/template), so a recipe is
```

> effectively a template that produces YAML. A recipe has two parts: a header

Reviewer: ```suggestion
effectively a template that produces YAML.

 A recipe has two parts:
 - a header
 - a list of actions.
 - 
 
 Recipe comments should start with `#`.
```

and a list of actions. Comments start with `#`.

> ## HEADER

Reviewer: ```suggestion
## RECIPE HEADER
```

> The header contains the following properties:

Reviewer: ```suggestion
The recipe header may contain the following properties:
```

- `architecture` -- target architecture (string). Mandatory; debos aborts if

>   it is missing.

Reviewer: ```suggestion
  it is missing (e.g. `amd64` or `arm64`).
```

- `sectorsize` -- logical sector size in bytes (integer). Optional, defaults
  to `512`. Set this to `4096` for devices using 4K blocks such as some UFS or
  NVMe storage. It is consumed by the `image-partition` and `raw` actions.
- `actions` -- the list of actions to run. Mandatory; at least one action is
  required.

> ```yaml

Reviewer: Mention that this is a minimal example

architecture: arm64
sectorsize: 512

actions:
  - action: debootstrap
    suite: trixie
```

> ## TEMPLATING

Reviewer: ```suggestion
## RECIPE TEMPLATE VARIABLES
```

> Two kinds of variables exist and they use different syntax:

Reviewer: ```suggestion
Two kinds of variables are passed to the recipe and they use different syntax:
```

- Variables passed on the command line with `-t NAME:VALUE` are fields of the

>   template data object and are referenced as `{{ .NAME }}`.

Reviewer: ```suggestion
  template data object and are referenced in the recipe as `{{ .NAME }}`.
```

> - Recipe-local variables are declared inside the recipe with

Reviewer: ```suggestion
- Local recipe variables are declared inside the recipe with
```

  `{{ $name := "value" }}` and referenced as `{{ $name }}`.

The common idiom combines both to give a command-line variable a default:

```yaml
{{- $image := or .image "debian.tgz" -}}

architecture: arm64

actions:
  - action: pack
    file: {{ $image }}
```

> Here `debos recipe.yaml` produces `debian.tgz`, while

Reviewer: ```suggestion
Here `debos recipe.yaml` creates `debian.tgz`, while
```

> `debos -t image:custom.tgz recipe.yaml` produces `custom.tgz`.

Reviewer: ```suggestion
`debos -t image:custom.tgz recipe.yaml` creates `custom.tgz`.
```

> The `recipe` action passes variables down to an included recipe; the

Reviewer: Does it ? I thought we only pass variables defined by `debos -t name:val` as `.name`?

Author: Yes, it does. Just tried it myself here.
I thought this is expected?

> `architecture` of an included recipe must match the parent and is injected

Reviewer: ```suggestion
The `architecture` property of an included recipe must match the parent and is injected
```

newline too

automatically as a template variable.

> ## TEMPLATE FUNCTIONS

Reviewer: ```suggestion
## RECIPE TEMPLATE FUNCTIONS
```

In addition to the standard Go template functions and the
[slim-sprig](https://go-task.github.io/slim-sprig/) function library, the
following custom functions are available:

- `escape` -- shell-escape the argument, e.g. `{{ escape $var }}`.
- `uuid5` -- generate a deterministic version 5 UUID,
  `{{ uuid5 $namespace $data }}`. `$namespace` must be a valid UUID. Passing

>   fixed values produces a reproducible UUID, useful for `fsuuid`/`partuuid`.

Reviewer: ```suggestion
  fixed values produces a reproducible UUID, which may be useful for the `image-partition` action.
```

POSSIBLY a llink to the docs for image-partition

> - `sector` -- append an `s` suffix, e.g. `{{ sector 256 }}` becomes `256s`.

Reviewer: ```suggestion
- `sector` -- appends an `s` suffix to the argument, e.g. `{{ sector 256 }}` becomes `256s`.
```

  Deprecated: the sector size is unknown at templating time, so write the
  offset directly as `256s` and let the action resolve it from `sectorsize`.

> Use `--print-recipe` to print the recipe after templating, and `--dry-run`

Reviewer: ```suggestion
Use `--print-recipe` to print the recipe after templating and `--dry-run`
```

to compose and verify the recipe without performing any work.

> # ACTIONS

Reviewer: ```suggestion
## RECIPE ACTIONS
```

> Actions are executed in the order they are listed. Unless

Reviewer: ```suggestion
Actions are executed in the order they are listed in the recipe.

Unless `--disable-fakemachine` is given (or no fakemachine backend is available), all actions
```

`--disable-fakemachine` is given (or no backend is available), all actions

> run inside the fakemachine virtual machine. The actions are grouped below by

Reviewer: ```suggestion
ran inside the fakemachine virtual machine. The actions are grouped below by
```

Author: huh? why past tense?

purpose.

> Every action accepts two common properties in addition to the

Reviewer: ```suggestion
Each action accepts two common properties in addition to the
```

action-specific ones listed below:

- `action` -- string, mandatory. Selects the action to run; it must be one of

>   the names documented below.

Reviewer: ```suggestion
  the actions documented below.
```

- `description` -- string, optional. Free-text label shown in debos output
  for this action; defaults to the action name.

Only the action-specific properties are listed in each action's section.

The following actions construct the target root filesystem.

> ## DEBOOTSTRAP

Reviewer: ```suggestion
## DEBOOTSTRAP ACTION
```

and same for the remaining actions

| ARTIFACTDIR | yes                    | no                   | yes                 |
| ROOTDIR     | yes                    | no                   | no                  |
| IMAGEMNTDIR | yes (if image mounted) | no                   | no                  |
| IMAGE       | yes (if image exists)  | yes (if image exists)| no                  |

Example -- run a script from the recipe inside the chroot:

```yaml
- action: run
  chroot: true
  script: scripts/setup-user.sh
```

Example -- build a squashfs of the rootfs on the host:

```yaml
- action: run
  chroot: false
  command: mksquashfs ${ROOTDIR} ${ARTIFACTDIR}/root.squashfs -comp xz -e rwdata/
```

Example -- run several commands as a single script with `command: |`:

```yaml
- action: run
  chroot: true
  command: |
    apt-get update
    apt-get -y dist-upgrade
    rm -rf /var/lib/apt/lists/*
```

Example -- compress the finished image as a postprocessing step:

```yaml
{{- $image := or .image "debian.img" -}}
- action: run
  postprocess: true
  command: gzip -f {{ $image }}
```

# OTHER EXAMPLE RECIPES

The [doc/examples](doc/examples) directory contains a complete, runnable

> example recipe, `ospack-debian/ospack-debian.yaml`, which builds a minimal

Reviewer: ```suggestion
example recipe `ospack-debian/ospack-debian.yaml` which builds a minimal
```

Link?!?!

> Debian ospack tarball. It exercises `debootstrap` (minbase), `apt`, `run` (a

Reviewer: Remove everything from "It exercises ..." - I think that makes sense. No need to explain what the recipe does.

chroot script and a non-chroot command using `${ROOTDIR}`), `overlay` and
`pack`, and ships its scripts and overlays.

> Additional more detailed example recipes are stored under [debos-recipes](https://github.com/go-debos/debos-recipes).

Reviewer: ```suggestion
Further more detailed example recipes are stored under the [debos-recipes](https://github.com/go-debos/debos-recipes) repository.
```

# ENVIRONMENT VARIABLES

## HOST-PROPAGATED VARIABLES

debos reads a predefined list of environment variables from the host and
propagates them to the fakemachine build environment. The list is defined by
`environVars` in `cmd/debos/debos.go` and currently covers the proxy
variables documented at
https://wiki.archlinux.org/index.php/proxy_settings:

```
http_proxy, https_proxy, ftp_proxy, rsync_proxy, all_proxy, no_proxy
```

> For each name both the lower and upper case variant is probed on the host

Reviewer: ```suggestion
For each of the above variables, both the lower and upper case variant is probed on the host
```

> and, if found, propagated to fakemachine preserving the case. So if the host

Reviewer: ```suggestion
and, if found, propagated to fakemachine preserving the case. So, if the host
```

defines `HTTP_PROXY` and `no_proxy`, both are propagated as-is.

> ## OVERRIDING WITH -E / --ENVIRON-VAR

Reviewer: ```suggestion
## SETTING AND OVERRIDING VARIABLES
```

The `--environ-var` and `-e` options specify, overwrite or unset environment
variables for fakemachine:

```bash
debos -e ENVIRONVAR:VALUE ...
```

> To unset a variable (prevent it from being propagated to fakemachine) use the

Reviewer: ```suggestion
To unset a variable (e.g. prevent it from being propagated to fakemachine) use the
```

same syntax without a value. Multiple `-e` options may be given.

## PROXY CONFIGURATION CAVEATS

While the proxy variables are exported from the host to fakemachine, there
are two known sources of issues:

- `localhost` will not work from fakemachine. Use an address that is valid on
  your network. debos warns if a variable contains `localhost`, `127.0.0.1`
  or `::1`.
- Applications and scripts run inside fakemachine may expect differently
  named or differently cased proxy variables; check which variables they use.

## VARIABLES EXPOSED TO THE RUN ACTION

The `run` action exposes `RECIPEDIR`, `ARTIFACTDIR`, `ROOTDIR`,
`IMAGEMNTDIR` and `IMAGE` to non-chroot commands and scripts. Which variables
are set depends on `chroot` and `postprocess`; see the table in the *run*
action above.

> ## THE ORIGIN PROPERTY

Reviewer: ```suggestion
## ORIGINS
```

> Several actions (`overlay`, `raw`, `unpack`, `install-deb`) accept an

Reviewer: maybe remove the list of actions ?

`origin` property that selects where their input comes from. The following
origins are always available:

- `recipe` -- the directory the recipe is in.
- `filesystem` -- the target filesystem root. Set by `filesystem-deploy` and
  `ostree-deploy` to the mounted image after deployment.

> - `artifacts` -- the artifact directory (the default origin for `unpack`).

Reviewer: ```suggestion
- `artifacts` -- the artifact directory.
```

In addition, every `download` action registers a named origin (its `name`
property) that resolves to the downloaded file, or to the extraction

> directory when `unpack` is enabled. Origins are resolved at run time, so a

Reviewer: ```suggestion
directory when `unpack` is enabled. Origins are resolved at runtime, so a
```
